source/openpulse/openpulse/parser.py

```diff
@@ -0,0 +1,303 @@
+        self._contexts: List[List[ParserRuleContext]] = []
+        self._in_defcal = in_defcal
+
+    @contextmanager
+    def _push_context(self, ctx: ParserRuleContext):
+        self._contexts.append([ctx])
+        yield
+        self._contexts.pop()
+
+    @contextmanager
+    def _push_scope(self, ctx: ParserRuleContext):
+        self._contexts[-1].append(ctx)
+        yield
+        self._contexts[-1].pop()
+
+    def _current_context(self):
+        return self._contexts[-1]
+
+    def _current_scope(self):
+        return self._contexts[-1][-1]
+
+    def _current_base_scope(self):
+        return self._contexts[-1][0]
+
+    def _in_global_scope(self):
+        return len(self._contexts) == 1 and len(self._contexts[0]) == 1
+
+    def _in_gate(self):
+        return isinstance(self._current_base_scope(), openpulseParser.GateStatementContext)
+
+    def _in_subroutine(self):
+        return isinstance(self._current_base_scope(), openpulseParser.DefStatementContext)
+
+    def _in_loop(self):
+        return any(
+            isinstance(
+                scope, (openpulseParser.ForStatementContext, openpulseParser.WhileStatementContext)
+            )
+            for scope in reversed(self._current_context())
+        )
+
+    @span
+    def _visitPulseType(self, ctx: openpulseParser.ScalarTypeContext):
+        if ctx.WAVEFORM():
+            return openpulse_ast.WaveformType()
+        if ctx.PORT():
+            return openpulse_ast.PortType()
+        if ctx.FRAME():
+            return openpulse_ast.FrameType()
+
+    @span
+    def visitCalibrationBlock(self, ctx: openpulseParser.CalibrationBlockContext):
+        with self._push_context(ctx):
+            return openpulse_ast.CalibrationBlock(
+                body=[self.visit(statement) for statement in ctx.openpulseStatement()]
+            )
+
+    def visitScalarType(self, ctx: openpulseParser.ScalarTypeContext):
+        if ctx.WAVEFORM() or ctx.PORT() or ctx.FRAME():
+            return self._visitPulseType(ctx)
+        else:
+            return QASMNodeVisitor.visitScalarType(self, ctx)
+
+    def visitIndexOperator(self, ctx: openpulseParser.IndexOperatorContext):
+        if ctx.setExpression():
+            return self.visit(ctx.setExpression())
+
+        index_element = (
+            openpulseParser.ExpressionContext,
+            openpulseParser.RangeExpressionContext,
+        )
+
+        def predicate(child):
+            return isinstance(child, index_element)
+
+        return [self.visit(child) for child in ctx.getChildren(predicate=predicate)]
+
+    @span
+    def visitRangeExpression(self, ctx: openpulseParser.RangeExpressionContext):
+        # start, end and step are all optional as in [:]
+        # It could be [start:end] or [start:step:end]
+        start = None
+        end = None
+        step = None
+        colons_seen = 0
+
+        for child in ctx.getChildren():
+            if isinstance(child, openpulseParser.ExpressionContext):
```

Comment:
@jakelishman, I tried your [Meta hack](https://github.com/openqasm/openqasm/pull/352#discussion_r950325285) but it failed to pass the unit tests. Somehow the code generated by different versions of Antlr is significantly different; I am using version 4.9.2 which is the one used in OpenQASM git action. I tried different hacks but each of them only pass a subset of the unit tests.